### DISCUSSION TITLE
How to send raw encoded bytes for strings with SQL Server ?

### DISCUSSION THREAD

**Discussion by D00mSlayer** [sqlalchemy/sqlalchemy#10979] [Usage Questions]
**Obsrervation**

I am migrating from py2 to py3 in a recent project I want to achieve the same effect between the 2 stacks with respect to storing unicode (utf-8 encoded) data in VARCHAR columns of SQL Server.

**Previous Stack**

- Python 2.7.18
- mxODBC
- SQLAlchemy 1.3.17
- SQL Server 2019

**Current Stack**

- Python 3.10
- pyodbc 4.0.35
- SQLAlchemy 1.3.17
- SQL Server 2019

**Example**

`import sqlalchemy as sa
from sqlalchemy import event

meta = sa.MetaData()
tbl = sa.Table(
    "table",
    meta,
    sa.Column("id", sa.Integer, primary_key=True),
    sa.Column("name", sa.String(200, "SQL_Latin1_General_CP1_CI_AS")),
)

e = sa.create_engine(
    "mssql+pyodbc://...driver=ODBC+Driver+18+for+SQL+Server",
    echo=True,
)

with e.connect() as c:
    meta.drop_all(c)
    meta.create_all(c)
    
    @event.listens_for(e, 'before_cursor_execute')
    def receive_before_cursor_execute(conn, cursor, statement, params, context, executemany):
        import pyodbc
        input_sizes = []
        for each_param in params:
            if isinstance(each_param, str):
                input_sizes.append((pyodbc.SQL_VARCHAR, 0, 0))  # doing this to avoid NVARCHAR parameterized queries
            else:
                input_sizes.append(())
        cursor.setinputsizes(input_sizes)

    c.execute(tbl.insert(), {"name": "メインメニュー"})`

This would try to match with CP1252; however it would not find any close match and hence would replace all the japanese characters with **?**.
In py2, i was able to encode the input with utf-8 and pass it down to the database, which would represent like this **ãƒ¡ã‚¤ãƒ³ãƒ¡ãƒ‹ãƒ¥ãƒ¼**.

With py3, if i change the model to use sa.BINARY instead of sa.String, I am able to save the data the same way. However, when i retrieve this data, I see errors like **typeerror: string argument without an encoding**. 

**Expectations**
I am pinned to the stack that is mentioned above.
How do i get the original input back when i query the DB?

**Comment by zzzeek** [comment]
pyodbc should be applying utf-8 encoding if that's the character encoding of your connection.    Googling around it's not fully clear to me how to ensure the ODBC connection is using utf-8, CC @gordthompson  for advice on this.

Otherwise, if we want to use a CHAR type that's going to do the utf-8 encoding before going to pyodbc, we can do this and I'd go full UserDefinedType for this since we want to change the driver conversation, you should be able to control the whole thing; here's a full POC:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
from sqlalchemy.types import UserDefinedType

Base = declarative_base()


class StringAsBytes(UserDefinedType):
    cache_ok = True

    def get_col_spec(self, **kw):
        return "VARCHAR(200) COLLATE SQL_Latin1_General_CP1_CI_AS"

    def bind_processor(self, dialect):
        def process(value):
            if value is not None:
                value = value.encode("utf-8")
            return value

        return process

    def result_processor(self, dialect, coltype):
        return None

    def get_dbapi_type(self, dbapi):
        return (dbapi.SQL_VARCHAR, 0, 0)


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String(200).with_variant(StringAsBytes, "mssql"))


e = create_engine(
    "mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/test?driver=ODBC+Driver+18+for+SQL+Server&TrustServerCertificate=yes",
    echo=True,
)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

s.add_all([A(data="hello world") for i in range(10)])

s.commit()

```

we can see bytes going in to the types:

```
INSERT INTO a (data) OUTPUT inserted.id, inserted.id AS id__1 SELECT p0 FROM (VALUES (?, 0), (?, 1), (?, 2), (?, 3), (?, 4), (?, 5), (?, 6), (?, 7), (?, 8), (?, 9)) AS imp_sen(p0, sen_counter) ORDER BY sen_counter
2024-02-07 08:37:20,459 INFO sqlalchemy.engine.Engine [generated in 0.00011s (insertmanyvalues) 1/1 (ordered)] (b'hello world', b'hello world', b'hello world', b'hello world', b'hello world', b'hello world', b'hello world', b'hello world', b'hello world', b'hello world')
2024-02-07 08:37:20,473 INFO sqlalchemy.engine.Engine COMMIT

```


setting the type up as String in the Python model level with the custom type as a variant will also help not involve the type too much with regular Python expression operations, it will only get called in when it's time to send or receive data.

**Reply by gordthompson** [reply]
> Googling around it's not fully clear to me how to ensure the ODBC connection is using utf-8, CC @gordthompson for advice on this.

MS ODBC drivers don't have the notion of "connection encoding" like, for example, mysql with `charset=utf8mb4`. They have historically used UTF-16, and Wireshark shows that they still use that wire-side, even for UTF-8 columns (which themselves are a relatively recent development).

@zzzeek - Your POC code does push the bytes as intended:

```python
s.add_all([A(data="メインメニュー") for i in range(2)])

s.commit()
```

gives us

```
id  data
1   ãƒ¡ã‚¤ãƒ³ãƒ¡ãƒ‹ãƒ¥ãƒ¼
2   ãƒ¡ã‚¤ãƒ³ãƒ¡ãƒ‹ãƒ¥ãƒ¼
```

on the server, but unfortunately it doesn't round trip

```python
a1 = s.get(A, 1)

a1_data = a1.data
print(a1_data)  # ãƒ¡ã‚¤ãƒ³ãƒ¡ãƒ‹ãƒ¥ãƒ¼
print(type(a1_data))  # <class 'str'>
```

To fix it we need to encode (back to bytes) and then decode as utf8:

```python
a1_data_fixed = a1_data.encode("cp1252").decode("utf8")
print(a1_data_fixed)  # メインメニュー
```

**Reply by zzzeek** [reply]
> To fix it we need to encode (back to bytes) and then decode as utf8:

ha ha of course I knew that!  I was just testing you

**Reply by D00mSlayer** [reply]
I see that the custom type is inserting with the utf-8 encoding. Is it possible to also retrieve data back with @gordthompson solution? I have this column select throughout my app. Cannot cp1526 encode and decode everywhere. 

Also, with the input i have provided in the post, it is working fine. But a character like this does not work - '成'

**Reply by gordthompson** [reply]
> Is it possible to also retrieve data back with @gordthompson solution?

```python
    def result_processor(self, dialect, coltype):
        def process(value):
            if isinstance(value, str):
                return value.encode("cp1252").decode("utf8")
            else:
                return None
        return process
```

> But a character like this does not work - '成'

That character is UTF-8 encoded as E6 88 90, and 90 is an unused position in [Windows-1252](https://en.wikipedia.org/wiki/Windows-1252).

**Reply by D00mSlayer** [reply]
@gordthompson yeah we saw that x90 is undefined in page 1252. In Py2, we could do a json.dumps and it would send the unicode code point and retain the original character. How come this isn't working in Py3?

**Reply by CaselIT** [reply]
You are probably not encoding into cp1252 in python2.

is using bytes on python3 not a possibility?

**Reply by D00mSlayer** [reply]
@CaselIT in Py2 we are encoding unicode to utf-8 into the VARCHAR columns. We're pinned to VARCHAR.

**Reply by CaselIT** [reply]
I'm talking only on the python side here

**Comment by gordthompson** [comment]
If you want to use JSON encoding then you could do something like this:

```python
import json

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
from sqlalchemy.types import UserDefinedType

Base = declarative_base()


class StringAsJsonStr(UserDefinedType):
    cache_ok = True

    def get_col_spec(self, **kw):
        return "VARCHAR(200) COLLATE SQL_Latin1_General_CP1_CI_AS"

    def bind_processor(self, dialect):
        def process(value):
            if value is not None:
                value = json.dumps(value, ensure_ascii=True)
            return value

        return process

    def result_processor(self, dialect, coltype):
        def process(value):
            if value is not None:
                value = json.loads(value)
            return value

        return process

    def get_dbapi_type(self, dbapi):
        return (dbapi.SQL_VARCHAR, 0, 0)


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String(200).with_variant(StringAsJsonStr, "mssql"))


e = create_engine(
    "mssql+pyodbc://scott:tiger^5HHH@mssql_199",
    echo=True,
)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

s.add_all([A(data="メインメニュー"), A(data="成")])

s.commit()

a = s.get(A, 1)
a_data = a.data
print(a_data)  # メインメニュー

a = s.get(A, 2)
a_data = a.data
print(a_data)  # 成
```

**Comment by D00mSlayer** [comment]
Thanks for the inputs @gordthompson @zzzeek @CaselIT 

We resolved it with this:

```
def get_bytes(data):
    byte_list = []

    for i in data:
        try:
            byte_list.append(i.encode('1252'))
        except UnicodeEncodeError:
            byte_list.append(i.encode('latin1'))

    return b''.join(byte_list)

class StringAsBytes(UserDefinedType):
    cache_ok = True

    def get_col_spec(self, **kw):
        return "VARCHAR(max) COLLATE SQL_Latin1_General_CP1_CI_AS"

    def bind_processor(self, dialect):
        def process(value):
            if value is not None:
                value = value.encode('utf-8')
            return value
        return process

    def result_processor(self, dialect, coltype):
        def process(value):
            if value is not None:
                try:
                    value = value.encode('1252').decode('utf-8')
                except Exception:
                    value = get_bytes(value).decode('utf-8', 'replace')
            return value
        return process

    def get_dbapi_type(self, dbapi):
        return (dbapi.SQL_VARCHAR, 0, 0)

```

Additionally, before cursor execution, we are forcing bytes to bind to SQL_VARCHAR:
```
import pyodbc
from sqlalchemy import event
from sqlalchemy.engine.base import Engine

@event.listens_for(Engine, 'before_cursor_execute')
def receive_before_cursor_execute(conn, cursor, statement, params, context, executemany):
    input_sizes = []
    for each_param in params:
        if isinstance(each_param, str):
            input_sizes.append((pyodbc.SQL_VARCHAR, 0, 0))
        elif isinstance(each_param, bytes):
            input_sizes.append((pyodbc.SQL_VARCHAR, 0, 0))
        else:
            input_sizes.append(())
    cursor.setinputsizes(input_sizes)

```

This way we were able to save and retrieve the entire unicode points successfully in Py3.

**Reply by zzzeek** [reply]
we have an event hook that's specific for the setinputsizes step, this would prevent the double setinputsizes call (SQLAlchemy is already calling cursor.setinputsizes before you are calling it again there)  note you are also blowing away the other inputsizes that SQLAlchemy might be setting which can be important especially for numeric datatypes

https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.DialectEvents.do_setinputsizes


the "inputsizes" variable is a dictionary of `BindParameter` -> DBAPI type object (or tuple), you can modify the dictionary in place, and that will be converted to the appropriate structure before being passed to cursor.setinputsizes

**Reply by D00mSlayer** [reply]
Isn't that a default in SQLA 2+? I'm on 1.3.17.

**Reply by zzzeek** [reply]
that hook was added in SQLAlchemy 1.2.9

**Reply by D00mSlayer** [reply]
Oh! I am only interested in strings and bytes. Is there a way to default the rest?

**Reply by zzzeek** [reply]
yah just dont change them in the dictionary you get

if you try the hook and print/pdb that structure you'll see what it is

**Reply by D00mSlayer** [reply]
I am not able to get into the hook with below code:

@event.listens_for(Engine, 'do_setinputsizes')
def receive_do_setinputsizes(inputsizes, cursor, statement, parameters, context):

**Reply by zzzeek** [reply]
here's POC, run this

```py
import pyodbc

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.engine.base import Engine
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
from sqlalchemy.types import UserDefinedType

Base = declarative_base()


@event.listens_for(Engine, 'do_setinputsizes')
def receive_do_setinputsizes(inputsizes, cursor, statement, parameters, context):
    for k, v in inputsizes.items():
        if isinstance(v, tuple) and v[0] == pyodbc.SQL_WVARCHAR:
            inputsizes[k] = (pyodbc.SQL_VARCHAR, 0, 0)


class StringAsBytes(UserDefinedType):
    cache_ok = True

    def get_col_spec(self, **kw):
        return "VARCHAR(200) COLLATE SQL_Latin1_General_CP1_CI_AS"

    def bind_processor(self, dialect):
        def process(value):
            if value is not None:
                value = value.encode("utf-8")
            return value

        return process

    def result_processor(self, dialect, coltype):
        return None

    def get_dbapi_type(self, dbapi):
        return (dbapi.SQL_VARCHAR, 0, 0)


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String(200).with_variant(StringAsBytes, "mssql"))


e = create_engine(
    "mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/test?driver=ODBC+Driver+18+for+SQL+Server&TrustServerCertificate=yes",
    echo=True,
)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

s.add_all([A(data="hello world") for i in range(10)])

s.commit()

```
